### DISCUSSION TITLE
DietPi-Software | Mosh

### DISCUSSION THREAD

**Discussion by hanscees** [MichaIng/DietPi#6097] [Software Request]
#### Is your feature request related to a problem? Please describe:
no
- ...

#### Describe the solution you'd like:
inclusion of the mosh shell which makes remote shell 1000x better 
https://mosh.org/
- ...

#### Describe alternatives you've considered:
mosh improves on ssh as remote shell, but takes over the connection via udp 
- ...

#### Additional context
see https://mosh.org/ 
its very beautiful with clients on for instance Ipad called blink 
https://blink.sh/

- ...

**Comment by MichaIng** [comment]
Thanks for your request.

Is this wrapped into a regular SSH session (detaching like GNU Screen does) or some own server-client model or how does it work?

EDIT: Converted to discussion, so people can now vote via UP button at your post.

**Comment by hanscees** [comment]
Hi, a good explanation begins here 
https://linuxhandbook.com/mosh/

Ssh server hands over the connection start to mosh server. Mosh server than uses port 60000-61000 UDP or something to do its thing. I do not know how this handing over of the connection is done exactly. 
OH wait, I do now: https://www.engagespark.com/blog/mosh-through-firewalls/
Some more of that is explained here 
https://stackoverflow.com/questions/61537871/using-mosh-instead-of-ssh

**Reply by MichaIng** [reply]
Okay got it:
- Authentication is done via SSH.
- After authentication, a communication on mentioned UDP ports is done between Mosh client and server.
- Those ports hence need to be forwarded by your router/NAT and allowed in firewall.
- On the server you hence need to run an SSH server as well as a Mosh server, and on the client you need to have an SSH client and a Mosh client. Mosh assumes some defaults for the SSH client, but in case you need to tell it which SSH executable, port and user to use.

For an implementation our end I'm not sure what we can do, since most setup steps need to be done on the client systems. On the server it is first of all just:
```sh
apt install mosh
```
For this alone an implementation wouldn't be worth it 😄. Let me know if there are reasonable setup steps we can do.

**Reply by georgimkv** [reply]
I think a reasonable setup step on the distro's end would be to open the default ports that Mosh uses. What do you think?

**Reply by Joulinar** [reply]
What do you mean to open ports? Where? On the router?
